config/src/runtime_upgrades.rs

```diff
@@ -0,0 +1,265 @@
+use std::{
+    fmt::Debug,
+    sync::{MutexGuard, PoisonError},
+};
+
+pub use serde::{Deserialize, Serialize};
+use thiserror::*;
+
+type Result<T, E = ReloadError> = std::result::Result<T, E>;
+
+/// Error which occurs when reloading a configuration fails.
+#[derive(Clone, Copy, Debug, Error)]
+pub enum ReloadError {
+    /// Locking the Mutex within [`ReloadHandle`] failed.  This error
+    /// should not occur frequently, and it is
+    #[error("Cannot acquire Mutex Lock.")]
+    PoisonedMutex,
+
+    /// If the reload handle wasn't properly initialised (using
+    /// [`ReloadHandle::set`]), there's nothing to reaload with.
+    #[error("Cannot reload an uninitialised handle.")]
+    NotInitialised,
+
+    /// Error not specified by the implementors of the [`Reloadable`]
+    /// traits. Use as last resort.
+    #[error("Unspecified reload failure.")]
+    Other,
+}
+
+type PoisonErr<'a, T> = PoisonError<MutexGuard<'a, Option<Box<(dyn ReloadMut<T> + Send + Sync)>>>>;
+
+impl<T> From<PoisonErr<'_, T>> for ReloadError {
+    fn from(_: PoisonErr<'_, T>) -> Self {
+        Self::PoisonedMutex
+    }
+}
+
+/// Implement this trait if the field must be mutably borrowed for `reload`.
+pub trait ReloadMut<T>: Debug {
+    // TODO: When negative traits/specialisation stabilises, Debug ought not to be a requirement.
+
+    /// Reload `self` using provided `item`.
+    ///
+    /// # Errors
+    /// Fails with an appropriate variant of
+    /// [`ReloadFailure`]. [`ReloadFailure::Other`] can be used as a
+    /// **temporary** placeholder.
+    fn reload(&mut self, item: T) -> Result<()>;
+}
+
+/// Implement this trait if the object can be immutably borrowed and `reload`-ed.
+pub trait Reload<T> {
+    /// Reload `self` using provided `item`.
+    ///
+    /// # Errors
+    /// Fails with an appropriate variant of [`ReloadFailure`].
+    /// [`ReloadFailure::Other`] can be used as a **temporary** placeholder.
+    fn reload(&self, item: T) -> Result<()>;
+}
+
+/// Contains [`handle`] types: opaque wrappers around a reloadable
+/// configuration.  [`handle::Reconfigurable`] is the most commonly
+/// useful type of handle, as it deserializes into the inner object,
+/// as well as handles synchronisation of changes to the program state
+/// via the handle, to the changes in the configuration value.
+pub mod handle {
```

Comment:
I think it would be very useful to have doc comments with minimal usage examples of each `Atomic`, `Singleton`, `Reconfigurable`.

Reply:
Also their naming is a bit inconsistent with each other to my mind.